### ISSUE TITLE
Wrong Code Similar to Pattern Matching on a Pair Raises an Internal Error

### ISSUE THREAD

**Issue by AliasQli** [idris-lang/Idris2#1441]
# Steps to Reproduce

```idris
module Main

(a, b) = (1, 2)
```

# Expected Behavior

Idris gives an error message about where and why it's wrong.

# Observed Behavior

```
Error: INTERNAL ERROR: (|((Builtin.Pair a) b),((Builtin.MkPair a) b)|) is not a function application
```

**Comment by gallais** [comment]
Similarly to #1216: The fix should be fairly similar to that of #1035 so I'll mark this as a good first issue.

**Comment by eayus** [comment]
Note that other expressions have the same problem too. For example,
```idris
module Main

case () of { () => () } = 5
```
also produces an internal error.

Instinctively I feel like the parser should be the thing rejecting these programs, is there a reason that is not the case?
